code/modules/reagents/chemistry/machinery/chem_dispenser.dm

```diff
@@ -395,3 +394,216 @@
+	"copper", "mercury", "plasma", "radium", "water", "ethanol", "sugar", "iodine", "bromine", "silver", "chromium")
+	var/current_reagent = null
+	var/efficiency = 0.2
+	var/recharge_rate = 1 // Keep this as an integer
+
+/obj/item/handheld_chem_dispenser/New()
+	..()
+	cell = new(src)
+	dispensable_reagents = sortList(dispensable_reagents)
+	current_reagent = pick(dispensable_reagents)
+	update_icon()
+	START_PROCESSING(SSobj, src)
+
+/obj/item/handheld_chem_dispenser/Destroy()
+	STOP_PROCESSING(SSobj, src)
+	return ..()
+
+/obj/item/handheld_chem_dispenser/get_cell()
+	return cell
+
+/obj/item/handheld_chem_dispenser/afterattack(obj/target, mob/user, proximity)
+	if(!proximity || !current_reagent || !amount)
+		return
+
+	if(!check_allowed_items(target,target_self = TRUE) || !target.is_refillable())
+		return
+
+	if(mode == "dispense")
+		var/free = target.reagents.maximum_volume - target.reagents.total_volume
+		var/actual = min(amount, cell.charge / efficiency, free)
+		target.reagents.add_reagent(current_reagent, actual)
+		cell.charge -= actual / efficiency
+		if(actual)
+			to_chat(user, "<span class='notice'>You dispense [amount] units of [current_reagent] into the [target].</span>")
+		update_icon()
+	else if(mode == "remove")
+		if(target.reagents.remove_reagent(current_reagent, amount))
+			to_chat(user, "<span class='notice'>You remove [amount] units of [current_reagent] from the [target].</span>")
+	else if(mode == "isolate")
+		if(target.reagents.isolate_reagent(current_reagent))
+			to_chat(user, "<span class='notice'>You remove all but [current_reagent] from the [target].</span>")
+
+/obj/item/handheld_chem_dispenser/attack_self(mob/user)
+	if(cell)
+		tgui_interact(user)
+	else
+		to_chat(user, "<span class='warning'>The [src] lacks a power cell!</span>")
+
+
+/obj/item/handheld_chem_dispenser/tgui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/tgui_state/state = GLOB.tgui_inventory_state)
+	// update the ui if it exists, returns null if no ui is passed/found
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "HandheldChemDispenser", ui_title, 390, 500)
+		ui.open()
+
+/obj/item/handheld_chem_dispenser/tgui_data(mob/user)
+	var/list/data = list()
+
+	data["glass"] = is_drink
+	data["amount"] = amount
+	data["energy"] = cell.charge ? cell.charge * efficiency : "0" //To prevent NaN in the UI.
+	data["maxEnergy"] = cell.maxcharge * efficiency
+	data["current_reagent"] = current_reagent
+	data["mode"] = mode
+
+	var/list/chemicals = list()
+	for(var/re in dispensable_reagents)
+		var/datum/reagent/temp = GLOB.chemical_reagents_list[re]
+		if(temp)
+			chemicals.Add(list(list("title" = temp.name, "id" = temp.id, "commands" = list("dispense" = temp.id)))) // list in a list because Byond merges the first list...
+	data["chemicals"] = chemicals
```

Comment:
Send this in `tgui_static_data()` to avoid excess sending of a list which wont change while the UI is open